File: annotation-processor/src/test/java/com/mayankrastogi/cs474/hw2/annotations/processor/tests/IteratorAnnotationProcessorTest.java
package com.mayankrastogi.cs474.hw2.annotations.processor.tests;

import com.google.testing.compile.Compilation;
import com.google.testing.compile.JavaFileObjects;
import com.mayankrastogi.cs474.hw2.annotations.processor.IteratorAnnotationProcessor;
import org.junit.Test;

import static com.google.testing.compile.CompilationSubject.assertThat;
import static com.google.testing.compile.Compiler.javac;

public class IteratorAnnotationProcessorTest {

    @Test
    public void Iterator_SucceedsOnValidUsage() {
        assertThat(compile("ValidIterator.java")).succeededWithoutWarnings();
    }

    @Test
    public void Iterator_CannotBeAppliedOnInterface() {
        assertThat(compile("IteratorAppliedOnInterface.java")).failed();
    }

    @Test
    public void Iterator_CanContainOnlyOneCurrentItemMethod() {
        assertThat(compile("IteratorWithNoCurrentItemMethod.java")).failed();
        assertThat(compile("IteratorWithTwoCurrentItemMethod.java")).failed();
    }

    @Test
    public void Iterator_CanContainOnlyOneIsDoneMethod() {
        assertThat(compile("IteratorWithNoIsDoneMethod.java")).failed();
        assertThat(compile("IteratorWithTwoIsDoneMethod.java")).failed();
    }

    @Test
    public void Iterator_CanContainOnlyOneNextItemMethod() {
        assertThat(compile("IteratorWithNoNextItemMethod.java")).failed();
        assertThat(compile("IteratorWithTwoNextItemMethod.java")).failed();
    }

    @Test
    public void CurrentItem_MustBeEnclosedWithinIterator() {
        assertThat(compile("CurrentItemNotInIterator.java")).failed();
    }

    @Test
    public void CurrentItem_ReturnTypeMustMatchIteratorAnnotationValue() {
        assertThat(compile("CurrentItemReturnTypeDifferentFromIteratorAnnotationValue.java")).failed();
    }

    @Test
    public void CurrentItem_CannotTakeAnyParameters() {
        assertThat(compile("CurrentItemWithParameters.java")).failed();
    }

    @Test
    public void IsDone_MustBeEnclosedWithinIterator() {
        assertThat(compile("IsDoneNotInIterator.java")).failed();
    }

    @Test
    public void IsDone_ReturnTypeMustBeBoolean() {
        assertThat(compile("IsDoneReturnTypeNotBoolean.java")).failed();
    }

    @Test
    public void IsDone_CannotTakeAnyParameters() {
        assertThat(compile("IsDoneWithParameters.java")).failed();
    }

    @Test
    public void IsDone_MustWarnIfPrivate() {
        assertThat(compile("IsDonePrivate.java")).hadWarningCount(1);
    }

    @Test
    public void NextItem_MustBeEnclosedWithinIterator() {
        assertThat(compile("NextItemNotInIterator.java")).failed();
    }

    @Test
    public void NextItem_ReturnTypeMustMatchIteratorAnnotationValue() {
        assertThat(compile("NextItemReturnTypeDifferentFromIteratorAnnotationValue.java")).failed();
    }

    @Test
    public void NextItem_CannotTakeAnyParameters() {
        assertThat(compile("NextItemWithParameters.java")).failed();
    }

    @Test
    public void NextItem_MustWarnIfPrivate() {
        assertThat(compile("NextItemPrivate.java")).hadWarningCount(1);
    }

    private Compilation compile(String fileName) {
        return javac()
                .withProcessors(new IteratorAnnotationProcessor())
                .compile(JavaFileObjects.forResource(fileName));
    }
}

File: annotation-processor/src/test/scala/com/mayankrastogi/cs474/hw2/annotations/processor/tests/IteratorAnnotationProcessorTest.scala
package com.mayankrastogi.cs474.hw2.annotations.processor.tests

import com.google.testing.compile.CompilationSubject.assertThat
import com.google.testing.compile.Compiler.javac
import com.google.testing.compile.{Compilation, JavaFileObjects}
import com.mayankrastogi.cs474.hw2.annotations.processor.IteratorAnnotationProcessor
import org.junit.runner.RunWith
import org.scalatest.FunSuite
import org.scalatestplus.junit.JUnitRunner

/**
 * Tests that the [[IteratorAnnotationProcessor]] is able to catch issues mentioned in this test suite.
 */
@RunWith(classOf[JUnitRunner])
class IteratorAnnotationProcessorTest extends FunSuite {

  test("IteratorAnnotationProcessor should succeed without warnings on valid usage.") {
    assertThat(compile("ValidIterator.java")).succeededWithoutWarnings()
  }

  test("@Iterator must not be applicable on an interface.") {
    assertThat(compile("IteratorAppliedOnInterface.java")).failed()
  }

  test("@Iterator must contain only one method annotated with @CurrentItem.") {
    assertThat(compile("IteratorWithNoCurrentItemMethod.java")).failed()
    assertThat(compile("IteratorWithTwoCurrentItemMethod.java")).failed()
  }

  test("@Iterator must contain only one method annotated with @IsDone.") {
    assertThat(compile("IteratorWithNoIsDoneMethod.java")).failed()
    assertThat(compile("IteratorWithTwoIsDoneMethod.java")).failed()
  }

  test("@Iterator must contain only one method annotated with @NextItem.") {
    assertThat(compile("IteratorWithNoNextItemMethod.java")).failed()
    assertThat(compile("IteratorWithTwoNextItemMethod.java")).failed()
  }

  test("@CurrentItem must be enclosed within a class annotated with @Iterator.") {
    assertThat(compile("CurrentItemNotInIterator.java")).failed()
  }

  test("Return type of method annotated with @CurrentItem must match the value property of its parent " +
    "class's @Iterator annotation.") {
    assertThat(compile("CurrentItemReturnTypeDifferentFromIteratorAnnotationValue.java")).failed()
  }

  test("Method annotated with @CurrentItem must not take any parameters.") {
    assertThat(compile("CurrentItemWithParameters.java")).failed()
  }

  test("@IsDone must be enclosed within a class annotated with @Iterator") {
    assertThat(compile("IsDoneNotInIterator.java")).failed()
  }

  test("Return type of method annotated with @IsDone must be boolean.") {
    assertThat(compile("IsDoneReturnTypeNotBoolean.java")).failed()
  }

  test("Method annotated with @IsDone must not take any parameters.") {
    assertThat(compile("IsDoneWithParameters.java")).failed()
  }

  test("Method annotated with @IsDone must generate a warning if it is private") {
    assertThat(compile("IsDonePrivate.java")).hadWarningCount(1)
  }

  test("@NextItem must be enclosed within a class annotated with @Iterator") {
    assertThat(compile("NextItemNotInIterator.java")).failed()
  }

  test("Return type of method annotated with @NextItem must match the value property of its parent " +
    "class's @Iterator annotation.") {
    assertThat(compile("NextItemReturnTypeDifferentFromIteratorAnnotationValue.java")).failed()
  }

  test("Method annotated with @NextItem must not take any parameters.") {
    assertThat(compile("NextItemWithParameters.java")).failed()
  }

  test("Method annotated with @NextItem must generate a warning if it is private") {
    assertThat(compile("NextItemPrivate.java")).hadWarningCount(1)
  }

  /**
   * Compiles the specified Java source file, present in the `src/test/resources` directory, and runs the
   * [[IteratorAnnotationProcessor]] during the compilation.
   *
   * @param fileName The Java source file to compile. The file should be present in the `src/test/resources` directory.
   * @return The result of the compilation.
   */
  private def compile(fileName: String): Compilation = {
    javac
      .withProcessors(new IteratorAnnotationProcessor)
      .compile(JavaFileObjects.forResource(fileName))
  }
}
